### PR TITLE
Fix `generate_index` ignoring default gems

### PR DESCRIPTION
**Pull request by juniorz**
`gem generate_index` should not ignore default gems if they are present in the directory containing the gems you want to index (the path pointed by the param --directory).

```
$ ruby -v && gem -v
ruby 2.1.2p95 (2014-05-08 revision 45877) [x86_64-darwin13.0]
2.4.5

$ tree .
.
└── gems
    ├── json-1.8.0.gem
    ├── json-1.8.1.gem
    └── sdoc-0.4.1.gem

1 directory, 3 files
$ gem generate_index -V
Generating Marshal quick index gemspecs for 2 gems
1/2: json-1.8.0
2/2: sdoc-0.4.1
Complete
Generated Marshal quick index gemspecs: 0.001s
Generating specs index
Generated specs index: 0.000s
Generating latest specs index
Generated latest specs index: 0.000s
Generating prerelease specs index
Generated prerelease specs index: 0.000s
Compressing indicies
Compressed indicies: 0.001s
Moving index into production dir .
mkdir -p ./quick
rm -rf ./quick/Marshal.4.8
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/quick/Marshal.4.8 ./quick/Marshal.4.8
rm -rf ./specs.4.8
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/specs.4.8 .
rm -rf ./specs.4.8.gz
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/specs.4.8.gz .
rm -rf ./latest_specs.4.8
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/latest_specs.4.8 .
rm -rf ./latest_specs.4.8.gz
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/latest_specs.4.8.gz .
rm -rf ./prerelease_specs.4.8
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/prerelease_specs.4.8 .
rm -rf ./prerelease_specs.4.8.gz
mv -f /var/folders/gp/2qhc4gmx08s3twkvzm9jbmdh0000gn/T/gem_generate_index_88548/prerelease_specs.4.8.gz .
```

**The problem:** `json-1.8.1.gem` is included in the `gems` directory but is simply ignored.
**The reason:** lib/rubygems/indexer.rb simply ignores every `default_gem?` regardless of its presence in the `gems` directory.

See https://github.com/rubygems/rubygems/issues/661#issuecomment-70561849 for more details.

### CODE REVIEW

**Comment by zzak** [comment]
This patch looks like it will always ignore default gems?

File: lib/rubygems/indexer.rb [hunk 1]
```diff
@@ -135,7 +145,6 @@ def build_marshal_gemspecs
 
     Gem.time 'Generated Marshal quick index gemspecs' do
       Gem::Specification.each do |spec|
-        next if spec.default_gem?
```
Comment (juniorz) [inline lib/rubygems/indexer.rb]: @zzak I don't think so. I've tested this approach.

This was the previous patch behavior (#661) - in this line it naively skipped every default_gem regardless of its inclusion in the directory containing the gems you want to index (the path pointed by the param `--directory`).

This patch, simply initialize the indexer's working set (`Gem::Specification#@_all`) with an empty set. This way, it ignores every default gem unless it is included in the directory containing the gems you want to index.

Do you have any example in mind?

**Comment by tenderlove** [comment]
Can you rebase this and I'll merge it?  It shouldn't ignore default gems.

**Comment by juniorz** [comment]
Sure.

**Comment by juniorz** [comment]
From latest version, it skips the default gems (json-1.8.1, minitest-4.7.5 and rake-10.1.0)

```
$ ruby -v && gem -v
ruby 2.1.2p95 (2014-05-08 revision 45877) [x86_64-darwin13.0]
2.4.6
$ tree .
.
├── gems
│   ├── json-1.8.1.gem
│   ├── json-1.8.2.gem
│   ├── minitest-4.7.5.gem
│   ├── rake-10.1.0.gem
│   └── sdoc-0.4.1.gem
├── latest_specs.4.8
├── latest_specs.4.8.gz
├── prerelease_specs.4.8
├── prerelease_specs.4.8.gz
├── quick
│   └── Marshal.4.8
│       ├── json-1.8.2.gemspec.rz
│       └── sdoc-0.4.1.gemspec.rz
├── specs.4.8
└── specs.4.8.gz

3 directories, 13 files
$ ruby -e 'p Gem::Specification.to_a.select(&:default_gem?).collect(&:full_name)'
["bigdecimal-1.2.4", "io-console-0.4.2", "json-1.8.1", "minitest-4.7.5", "psych-2.0.5", "rake-10.1.0", "rdoc-4.1.0", "test-unit-2.1.2.0"]
$ gem generate_index -V                                                                                           [⚡ ∞][15:39:21]
Generating Marshal quick index gemspecs for 2 gems
1/2: json-1.8.2
2/2: sdoc-0.4.1
Complete
[...]
```

Running from master, it doesn't.

```
$ ruby -I/Users/king/src/github/rubygems/lib -e 'p Gem::Specification.to_a.select(&:default_gem?).collect(&:full_name)'
["bigdecimal-1.2.4", "io-console-0.4.2", "json-1.8.1", "minitest-4.7.5", "psych-2.0.5", "rake-10.1.0", "rdoc-4.1.0", "test-unit-2.1.2.0"]
$ ruby -I/Users/king/src/github/rubygems/lib /Users/king/src/github/rubygems/bin/gem generate_index -v
Generating Marshal quick index gemspecs for 5 gems
1/5: json-1.8.1
2/5: json-1.8.2
3/5: minitest-4.7.5
4/5: rake-10.1.0
5/5: sdoc-0.4.1
Complete
[...]
```

From my perspective, it was decided to use `Gem::Specification.each` to iterate over all the gem specs. To do so, it had to be mutated in order to include the gems located in the destination directory. Then #661 was reported because `Gem::Specification.each` includes all the default gems, and it was fixed by skipping the default gems :-1:

Then this issue was reported because the default gems are skipped even if they are not coming from the default specifications directory (`Gem::BasicSpecification.default_specifications_dir`).

aad1014e124d65ed28dcd6580cbf6cf8441c1690 removes mutation on `Gem::Specification` and by consequence fixes this issue.

That makes me wonder if the `next if spec.default_gem?` is really necessary since it seems to be originally added to fix a consequence of mutating `Gem:Specification` (see #661).

Now, `default_gem?` would only return true if `gem generate_index` is run on `Gem::BasicSpecification.default_specifications_dir` (the directory containing only the default gems) - and in this case it would skip all the gems.

That's why my PR becomes only removing the skip behavior.

**Comment by segiddins** [comment]
I'm worried that this doesn't break any tests -- from reading the prior comments, it seems this change should be good? Presumably we could merge with added test coverage?

**Comment by drbrain** [comment]
If we have tests for the scenarios:

* have default gem without matching .gem file
* have default gem with matching .gem file

Then we can merge this.

**Comment by copiousfreetime** [comment]
@juniorz looks like all we need is the couple of tests that @drbrain requests and we'll be good to go.

**Comment by juniorz** [comment]
I'll take a look.

**Comment by bronzdoc** [comment]
Closing since this has been stale for to long, please add the required tests to re-open and move forward with this PR, thanks.

**Pull request closed without merging**